ProcessInstanceVariableTest#shouldUpdateMultipleVariables is flaky

**Summary**

- How often does the test fail? 3 times so far
- Does it block your work? not particularly
- Do we suspect that it is a real failure? no idea

**Failures**

<details><summary>Example assertion failure</summary>
<pre>
Stacktrace
java.lang.AssertionError: 
Expecting ArrayList:
  [("x", "1"), ("z", "5")]
to contain:
  [("y", "4"), ("z", "5")]
but could not find the following element(s):
  [("y", "4")]
	at io.camunda.zeebe.engine.processing.variable.ProcessInstanceVariableTest.shouldUpdateMultipleVariables(ProcessInstanceVariableTest.java:321)
</pre>
</details>

**Logs**

<details><summary>Logs</summary>

```
Standard Output
12:29:41.120 [] INFO  io.camunda.zeebe.test - Test failed, following records were exported:
12:29:41.255 [] INFO  io.camunda.zeebe.test - Compact log representation:
--------
	['C'ommand/'E'event/'R'ejection] [valueType] [intent] - #[position]->#[source record position]  P[partitionId]K[key] - [summary of value]
	P9K999 - key; #999 - record position; "ID" element/process id; @"elementid"/[P9K999] - element with ID and key
	Keys are decomposed into partition id and per partition key (e.g. 2251799813685253 -> P1K005). If single partition, the partition is omitted.
	Long IDs are shortened (e.g. 'startEvent_5d56488e-0570-416c-ba2d-36d2a3acea78' -> 'star..acea78'
--------
E PROC_INST          COMPLETING     - #36->#35 K15 - START_EVENT "startEv..65e9fab" in <process "process"[K11]>
E PROC_INST          COMPLETED      - #37->#35 K15 - START_EVENT "startEv..65e9fab" in <process "process"[K11]>
E PROC_INST          SEQ_FLOW_TAKEN - #38->#35 K16 - SEQUENCE_FLOW "sequenc..7a1b671" in <process "process"[K11]>
C PROC_INST          ACTIVATE       - #39->#35 K17 - SERVICE_TASK "task" in <process "process"[K11]>
C PROC_INST_CREATION CREATE         - #40-> -1  -1 - new <process "process"> with variables: {x=1, y=2, z=3}
E PROC_INST          ACTIVATING     - #41->#39 K17 - SERVICE_TASK "task" in <process "process"[K11]>
E JOB                CREATED        - #42->#39 K18 - K18 "test" @"task"[K17] 3 retries, in <process "process"[K11]> (no vars)
E PROC_INST          ACTIVATED      - #43->#39 K17 - SERVICE_TASK "task" in <process "process"[K11]>
E VAR                CREATED        - #44->#40 K20 - x->1 in <process [K19]>
E VAR                CREATED        - #45->#40 K21 - z->3 in <process [K19]>
E VAR                CREATED        - #46->#40 K22 - y->2 in <process [K19]>
C PROC_INST          ACTIVATE       - #47->#40 K19 - PROCESS "process" in <process "process"[K19]>
E PROC_INST_CREATION CREATED        - #48->#40 K23 - new <process "process"> with variables: {x=1, y=2, z=3}
E PROC_INST          ACTIVATING     - #49->#47 K19 - PROCESS "process" in <process "process"[K19]>
E PROC_INST          ACTIVATED      - #50->#47 K19 - PROCESS "process" in <process "process"[K19]>
C PROC_INST          ACTIVATE       - #51->#47  -1 - START_EVENT "startEv..65e9fab" in <process "process"[K19]>
E PROC_INST          ACTIVATING     - #52->#51 K24 - START_EVENT "startEv..65e9fab" in <process "process"[K19]>
E PROC_INST          ACTIVATED      - #53->#51 K24 - START_EVENT "startEv..65e9fab" in <process "process"[K19]>
C PROC_INST          COMPLETE       - #54->#51 K24 - START_EVENT "startEv..65e9fab" in <process "process"[K19]>
E PROC_INST          COMPLETING     - #55->#54 K24 - START_EVENT "startEv..65e9fab" in <process "process"[K19]>
E PROC_INST          COMPLETED      - #56->#54 K24 - START_EVENT "startEv..65e9fab" in <process "process"[K19]>
E PROC_INST          SEQ_FLOW_TAKEN - #57->#54 K25 - SEQUENCE_FLOW "sequenc..7a1b671" in <process "process"[K19]>
C PROC_INST          ACTIVATE       - #58->#54 K26 - SERVICE_TASK "task" in <process "process"[K19]>
E PROC_INST          ACTIVATING     - #59->#58 K26 - SERVICE_TASK "task" in <process "process"[K19]>
E JOB                CREATED        - #60->#58 K27 - K27 "test" @"task"[K26] 3 retries, in <process "process"[K19]> (no vars)
E PROC_INST          ACTIVATED      - #61->#58 K26 - SERVICE_TASK "task" in <process "process"[K19]>
C JOB                COMPLETE       - #62-> -1 K27 - K27 "test" @""[ -1] -1 retries, in <process ?[?]> with variables: {x=1, y=4, z=5}
E JOB                COMPLETED      - #63->#62 K27 - K27 "test" @"task"[K26] 3 retries, in <process "process"[K19]> with variables: {x=1, y=4, z=5}
C PROC_INST          COMPLETE       - #64->#62 K26 - SERVICE_TASK "task" in <process "process"[K19]>
E PROC_INST          COMPLETING     - #65->#64 K26 - SERVICE_TASK "task" in <process "process"[K19]>
E VAR                UPDATED        - #66->#64 K20 - x->1 in <process [K19]>
E VAR                UPDATED        - #67->#64 K21 - z->5 in <process [K19]>
E VAR                UPDATED        - #68->#64 K22 - y->4 in <process [K19]>
E PROC_INST          COMPLETED      - #69->#64 K26 - SERVICE_TASK "task" in <process "process"[K19]>
E PROC_INST          SEQ_FLOW_TAKEN - #70->#64 K28 - SEQUENCE_FLOW "sequenc..f9809c2" in <process "process"[K19]>
C PROC_INST          ACTIVATE       - #71->#64 K29 - END_EVENT "endEven..c88755c" in <process "process"[K19]>
E PROC_INST          ACTIVATING     - #72->#71 K29 - END_EVENT "endEven..c88755c" in <process "process"[K19]>
E PROC_INST          ACTIVATED      - #73->#71 K29 - END_EVENT "endEven..c88755c" in <process "process"[K19]>
E PROC_INST          COMPLETING     - #74->#71 K29 - END_EVENT "endEven..c88755c" in <process "process"[K19]>
E PROC_INST          COMPLETED      - #75->#71 K29 - END_EVENT "endEven..c88755c" in <process "process"[K19]>
C PROC_INST          COMPLETE       - #76->#71 K19 - PROCESS "process" in <process "process"[K19]>
E PROC_INST          COMPLETING     - #77->#76 K19 - PROCESS "process" in <process "process"[K19]>
E PROC_INST          COMPLETED      - #78->#76 K19 - PROCESS "process" in <process "process"[K19]>
-------------- Deployed Processes ----------------------
--------------- Decomposed keys (for debugging) -----------------
 -1 <-> -1
K11 <-> 2251799813685259
K15 <-> 2251799813685263
K16 <-> 2251799813685264
K17 <-> 2251799813685265
K18 <-> 2251799813685266
K19 <-> 2251799813685267
K20 <-> 2251799813685268
K21 <-> 2251799813685269
K22 <-> 2251799813685270
K23 <-> 2251799813685271
K24 <-> 2251799813685272
K25 <-> 2251799813685273
K26 <-> 2251799813685274
K27 <-> 2251799813685275
K28 <-> 2251799813685276
K29 <-> 2251799813685277
```

</details>

--- a/engine/src/main/java/io/camunda/zeebe/engine/processing/variable/VariableBehavior.java
+++ b/engine/src/main/java/io/camunda/zeebe/engine/processing/variable/VariableBehavior.java
@@ -123,7 +123,7 @@ public final class VariableBehavior {
         final VariableInstance variableInstance =
             variableState.getVariableInstanceLocal(currentScope, entry.getName());
 
-        if (variableInstance != null) {
+        if (variableInstance != null && !variableInstance.getValue().equals(entry.getValue())) {
           applyEntryToRecord(entry);
           stateWriter.appendFollowUpEvent(
               variableInstance.getKey(), VariableIntent.UPDATED, variableRecord);
@@ -182,7 +182,7 @@ public final class VariableBehavior {
     if (variableInstance == null) {
       final long key = keyGenerator.nextKey();
       stateWriter.appendFollowUpEvent(key, VariableIntent.CREATED, record);
-    } else {
+    } else if (!variableInstance.getValue().equals(record.getValueBuffer())) {
       stateWriter.appendFollowUpEvent(variableInstance.getKey(), VariableIntent.UPDATED, record);
     }
   }

--- a/engine/src/test/java/io/camunda/zeebe/engine/processing/incident/MultiInstanceIncidentTest.java
+++ b/engine/src/test/java/io/camunda/zeebe/engine/processing/incident/MultiInstanceIncidentTest.java
@@ -33,7 +33,9 @@ import io.camunda.zeebe.test.util.record.RecordingExporterTestWatcher;
 import java.util.Arrays;
 import java.util.Collections;
 import java.util.List;
+import java.util.Map;
 import java.util.Set;
+import java.util.stream.Collectors;
 import org.junit.Before;
 import org.junit.ClassRule;
 import org.junit.Rule;
@@ -293,20 +295,24 @@ public final class MultiInstanceIncidentTest {
             .create();
 
     // when
-    RecordingExporter.incidentRecords(IncidentIntent.CREATED)
-        .withProcessInstanceKey(processInstanceKey)
-        .limit(3)
-        .map(Record::getKey)
+    final List<Long> incidents =
+        RecordingExporter.incidentRecords(IncidentIntent.CREATED)
+            .withProcessInstanceKey(processInstanceKey)
+            .limit(3)
+            .map(Record::getKey)
+            .collect(Collectors.toList());
+    ENGINE.variables().ofScope(processInstanceKey).withDocument(Map.of("y", 1)).update();
+    incidents.stream()
         .map(key -> ENGINE.incident().ofInstance(processInstanceKey).withKey(key))
         .forEach(ResolveIncidentClient::resolve);
 
     // then
     final var variableNames = Set.of("item", "loopCounter");
     assertThat(
-            RecordingExporter.variableRecords()
-                .withProcessInstanceKey(processInstanceKey)
-                .filter(v -> variableNames.contains(v.getValue().getName()))
-                .limit(12))
+            RecordingExporter.records()
+                .limitToProcessInstance(processInstanceKey)
+                .variableRecords()
+                .filter(v -> variableNames.contains(v.getValue().getName())))
         .extracting(v -> tuple(v.getIntent(), v.getValue().getName(), v.getValue().getValue()))
         .containsExactly(
             tuple(VariableIntent.CREATED, "item", "1"),
@@ -314,13 +320,7 @@ public final class MultiInstanceIncidentTest {
             tuple(VariableIntent.CREATED, "item", "2"),
             tuple(VariableIntent.CREATED, "loopCounter", "2"),
             tuple(VariableIntent.CREATED, "item", "3"),
-            tuple(VariableIntent.CREATED, "loopCounter", "3"),
-            tuple(VariableIntent.UPDATED, "item", "1"),
-            tuple(VariableIntent.UPDATED, "loopCounter", "1"),
-            tuple(VariableIntent.UPDATED, "item", "2"),
-            tuple(VariableIntent.UPDATED, "loopCounter", "2"),
-            tuple(VariableIntent.UPDATED, "item", "3"),
-            tuple(VariableIntent.UPDATED, "loopCounter", "3"));
+            tuple(VariableIntent.CREATED, "loopCounter", "3"));
   }
 
   @Test

--- a/engine/src/test/java/io/camunda/zeebe/engine/processing/variable/ProcessInstanceVariableTest.java
+++ b/engine/src/test/java/io/camunda/zeebe/engine/processing/variable/ProcessInstanceVariableTest.java
@@ -312,9 +312,11 @@ public final class ProcessInstanceVariableTest {
 
     // then
     assertThat(
-            RecordingExporter.variableRecords(VariableIntent.UPDATED)
-                .withProcessInstanceKey(processInstanceKey)
-                .limit(2))
+            RecordingExporter.records()
+                .limitToProcessInstance(processInstanceKey)
+                .variableRecords()
+                .withIntent(VariableIntent.UPDATED)
+                .withProcessInstanceKey(processInstanceKey))
         .extracting(Record::getValue)
         .extracting(v -> tuple(v.getName(), v.getValue()))
         .hasSize(2)

--- a/engine/src/test/java/io/camunda/zeebe/engine/processing/variable/VariableBehaviorTest.java
+++ b/engine/src/test/java/io/camunda/zeebe/engine/processing/variable/VariableBehaviorTest.java
@@ -194,6 +194,38 @@ final class VariableBehaviorTest {
   }
 
   @Test
+  void shouldMergeDocumentWithoutUpdatingUnmodifiedVariable() {
+    // given
+    final long processDefinitionKey = 1;
+    final long rootScopeKey = 1;
+    final long parentScopeKey = 2;
+    final long childScopeKey = 3;
+    final Map<String, Object> document = Map.of("foo", "bar", "buz", "baz");
+    state.createScope(rootScopeKey, VariableState.NO_PARENT);
+    state.createScope(parentScopeKey, rootScopeKey);
+    state.createScope(childScopeKey, parentScopeKey);
+    setVariable(4, rootScopeKey, processDefinitionKey, "foo", "bar");
+
+    // when
+    behavior.mergeDocument(
+        childScopeKey, processDefinitionKey, rootScopeKey, MsgPackUtil.asMsgPack(document));
+
+    // then
+    final List<RecordedEvent<VariableRecordValue>> events = getFollowUpEvents();
+    assertThat(events)
+        .satisfiesExactlyInAnyOrder(
+            event -> {
+              assertThat(event.intent).isEqualTo(VariableIntent.CREATED);
+              VariableRecordValueAssert.assertThat(event.value)
+                  .hasName("buz")
+                  .hasValue("\"baz\"")
+                  .hasScopeKey(rootScopeKey)
+                  .hasProcessDefinitionKey(processDefinitionKey)
+                  .hasProcessInstanceKey(rootScopeKey);
+            });
+  }
+
+  @Test
   void shouldMergeDocumentWithoutPropagatingExistingVariables() {
     // given
     final long processDefinitionKey = 1;
@@ -320,6 +352,34 @@ final class VariableBehaviorTest {
                   .hasProcessDefinitionKey(processDefinitionKey)
                   .hasProcessInstanceKey(parentScopeKey);
             });
+  }
+
+  @Test
+  void shouldOnlyUpdateModifiedVariables() {
+    // given
+    final long processDefinitionKey = 1;
+    final long parentScopeKey = 1;
+    final long childScopeKey = 2;
+    final long parentFooKey = 3;
+    final DirectBuffer variableName = BufferUtil.wrapString("foo");
+    final DirectBuffer variableValue = packString("bar");
+    state.createScope(parentScopeKey, VariableState.NO_PARENT);
+    state.createScope(childScopeKey, parentScopeKey);
+    setVariable(parentFooKey, parentScopeKey, processDefinitionKey, "foo", "bar");
+
+    // when
+    behavior.setLocalVariable(
+        parentScopeKey,
+        processDefinitionKey,
+        parentScopeKey,
+        variableName,
+        variableValue,
+        0,
+        variableValue.capacity());
+
+    // then
+    final List<RecordedEvent<VariableRecordValue>> events = getFollowUpEvents();
+    assertThat(events).isEmpty();
   }
 
   @SuppressWarnings("unchecked")

--- a/engine/src/test/java/io/camunda/zeebe/engine/processing/variable/VariableBehaviorTest.java
+++ b/engine/src/test/java/io/camunda/zeebe/engine/processing/variable/VariableBehaviorTest.java
@@ -355,7 +355,7 @@ final class VariableBehaviorTest {
   }
 
   @Test
-  void shouldOnlyUpdateModifiedVariables() {
+  void shouldNotUpdateUnmodifiedVariables() {
     // given
     final long processDefinitionKey = 1;
     final long parentScopeKey = 1;

--- a/engine/src/test/java/io/camunda/zeebe/engine/util/RecordingTypedEventWriter.java
+++ b/engine/src/test/java/io/camunda/zeebe/engine/util/RecordingTypedEventWriter.java
@@ -42,5 +42,10 @@ public final class RecordingTypedEventWriter implements TypedEventWriter {
       this.intent = intent;
       this.value = (T) Records.cloneValue(value);
     }
+
+    @Override
+    public String toString() {
+      return "RecordedEvent{" + "key=" + key + ", intent=" + intent + ", value=" + value + '}';
+    }
   }
 }

